src/java.base/share/classes/java/lang/template/TemplateRuntime.java

```diff
@@ -0,0 +1,162 @@
+    }
+
+    /**
+     * Generic StringTemplate.
+     *
+     * @param fragments  immutable list of string fragments from string template
+     * @param values     immutable list of expression values
+     */
+    record SimpleStringTemplate(List<String> fragments,
+                                 List<Object> values
+    ) implements StringTemplate {}
+
+    /**
+     * Creates a string that interleaves the elements of values between the
+     * elements of fragments.
+     *
+     * @param fragments  list of String fragments
+     * @param values     list of expression values
+     *
+     * @return String interpolation of fragments and values
+     */
+    static String interpolate(List<String> fragments, List<Object> values) {
+        int fragmentsSize = fragments.size();
+        int valuesSize = values.size();
+        if (fragmentsSize == 1) {
+            return fragments.get(0);
+        }
+        int size = fragmentsSize + valuesSize;
+        String[] strings = new String[size];
+        Iterator<String> fragmentsIter = fragments.iterator();
+        int i = 0;
+        for (Object value : values) {
+            strings[i++] = fragmentsIter.next();
+            strings[i++] = String.valueOf(value);
+        }
+        strings[i++] = fragmentsIter.next();
+        return JLA.join("", "", "", strings, size);
+    }
+
+    /**
+     * Combine one or more {@link StringTemplate StringTemplates} to produce a combined {@link StringTemplate}.
+     * {@snippet :
+     * StringTemplate st = StringTemplate.combine("\{a}", "\{b}", "\{c}");
+     * assert st.interpolate().equals("\{a}\{b}\{c}");
+     * }
+     *
+     * @param sts  one or more {@link StringTemplate}
+     *
+     * @return combined {@link StringTemplate}
+     *
+     * @throws NullPointerException if sts is null or if any element of sts is null
+     */
+    static StringTemplate combine(StringTemplate... sts) {
+        Objects.requireNonNull(sts, "sts must not be null");
+        if (sts.length == 0) {
+            return StringTemplate.of("");
+        } else if (sts.length == 1) {
+            return Objects.requireNonNull(sts[0], "string templates should not be null");
+        }
+        int size = 0;
+        for (StringTemplate st : sts) {
+            Objects.requireNonNull(st, "string templates should not be null");
+            size += st.values().size();
+        }
+        String[] fragments = new String[size + 1];
+        Object[] values = new Object[size];
+        int i = 0, j = 0;
+        fragments[0] = "";
+        for (StringTemplate st : sts) {
+            Iterator<String> fragmentIter = st.fragments().iterator();
+            fragments[i++] += fragmentIter.next();
+            while (fragmentIter.hasNext()) {
+                fragments[i++] = fragmentIter.next();
+            }
+            i--;
+            for (Object value : st.values()) {
+                values[j++] = value;
+            }
+        }
```

Comment:
The sides effects `fragments[i++] += fragmentIter.next();` and `i--` are ugly and can easily be overlook.
Also there is no need to use an iterator here, we know that the fragments have a List.get() in O(1).
The idea is to concat the last fragment of a template with the first one of the next template, so i propose
```java
        ...
        String[] fragments = new String[size + 1];
        String last = "";
        int i = 0, j = 0;
        for (StringTemplate st : sts) {
            List<String> templateFragments = st.fragments();
            fragments[i++] = last.concat(templateFragments.get(0));  // concat last fragment with first new fragment
            int k = 0;
            for(; k < templateFragments.size() - 1; k++) {
              fragments[i++] = templateFragments.get(k);
            }
            last = templateFragments.get(k);
            for (Object value : st.values()) {
                values[j++] = value;
            }
        }
        fragments[i] = last;
```